src/coreclr/jit/promotion.cpp

```diff
@@ -0,0 +1,1310 @@
+            size_t len  = strlen(buf) + 1;
+            char*  bufp = new (comp, CMK_DebugOnly) char[len];
+            strcpy_s(bufp, len, buf);
+#endif
+            unsigned   newLcl = comp->lvaGrabTemp(false DEBUGARG(bufp));
+            LclVarDsc* dsc    = comp->lvaGetDesc(newLcl);
+            dsc->lvType       = access.AccessType;
+
+            if (*replacements == nullptr)
+            {
+                *replacements =
+                    new (comp, CMK_Promotion) jitstd::vector<Replacement>(comp->getAllocator(CMK_Promotion));
+            }
+
+            (*replacements)->push_back(Replacement(access.Offset, access.AccessType, newLcl));
+        }
+    }
+
+    //------------------------------------------------------------------------
+    // EvaluateReplacement:
+    //   Evaluate legality and profitability of a single replacement candidate.
+    //
+    // Parameters:
+    //   comp   - Compiler instance
+    //   lclNum - Local num for this struct local
+    //   access - Access information for the candidate.
+    //
+    // Returns:
+    //   True if we should promote this access and create a replacement; otherwise false.
+    //
+    bool EvaluateReplacement(Compiler* comp, unsigned lclNum, const Access& access)
+    {
+        weight_t countOverlappedCallsWtd                 = 0;
+        weight_t countOverlappedReturnsWtd               = 0;
+        weight_t countOverlappedRetbufsWtd               = 0;
+        weight_t countOverlappedAssignmentDestinationWtd = 0;
+        weight_t countOverlappedAssignmentSourceWtd      = 0;
+
+        bool overlap = false;
+        for (const Access& otherAccess : m_accesses)
+        {
+            if (&otherAccess == &access)
+                continue;
+
+            if (!otherAccess.Overlaps(access.Offset, genTypeSize(access.AccessType)))
+            {
+                continue;
+            }
+
+            if (otherAccess.AccessType != TYP_STRUCT)
+            {
+                return false;
+            }
+
+            countOverlappedCallsWtd += otherAccess.CountCallArgsWtd;
+            countOverlappedReturnsWtd += otherAccess.CountReturnsWtd;
+            countOverlappedRetbufsWtd += otherAccess.CountPassedAsRetbufWtd;
+            countOverlappedAssignmentDestinationWtd += otherAccess.CountAssignmentDestinationWtd;
+            countOverlappedAssignmentSourceWtd += otherAccess.CountAssignmentSourceWtd;
+        }
+
+        // TODO-CQ: Tune the following heuristics. Currently they are based on
+        // x64 code size although using BB weights when available.
+        weight_t costWithout = 0;
+
+        // A normal access without promotion looks like:
+        // mov reg, [reg+offs]
+        // It may also be contained. Overall we are going to cost each use of
+        // an unpromoted local at 6.5 bytes.
+        // TODO-CQ: We can make much better guesses on what will and won't be contained.
+        costWithout += access.CountWtd * 6.5;
```

Comment:
This seems to be mixing units? Weighted costs reflect (relative) frequency so should be used to scale times, but bytes are size related.

You may want to form both a time score and a size score and then explicitly try and model the tradeoffs between the two.

Eg for inlining I would generally propose: if size is negative then always transform; else if size increases transform only if ratio of time improvement to size increase is above some threshold.

Reply:
Initially this was purely a size estimate but then I switched it to make some use of PGO weights. But I didn't really spend any time on the costing after that, it's pretty hard to make good choices when a lot of the basic stuff is still missing.
I will add a todo here -- your suggestion to eventually check both size and perf makes sense. But I found it very hard to predict the size cost (in particular due to prolog/epilog cost and potential spilling). And accurately modelling execution cost is also going to be pretty hard.

Reply:
Actually I can see there is already a TODO above for it:
```
// TODO-CQ: Tune the following heuristics. Currently they are based on
// x64 code size although using BB weights when available.
```

Reply:
I added clarification to the todo that this mixing of units does not make sense.